packages/protocolbeat/src/panel-terminal/TerminalPanel.tsx

```diff
@@ -0,0 +1,110 @@
+import { useQuery, useQueryClient } from '@tanstack/react-query'
+import { useEffect, useRef, useState } from 'react'
+import { useParams } from 'react-router-dom'
+import { executeCommand, getProject } from '../api/api'
+import { useTerminalStore } from './store'
+
+export function TerminalPanel() {
+  const queryClient = useQueryClient()
+  const { project } = useParams()
+  const [discoveryChain, setDiscoveryChain] = useState<string | undefined>(
+    undefined,
+  )
+  const abortControllerRef = useRef<AbortController>()
+  const outputRef = useRef<HTMLDivElement>(null)
+  const { output, isRunning, addOutput, setIsRunning, clear } =
+    useTerminalStore()
+
+  if (!project) {
+    throw new Error('Cannot use component outside of project page!')
+  }
+
+  const getProjectResponse = useQuery({
+    queryKey: ['projects', project],
+    queryFn: () => getProject(project),
+  })
+  const chains = getProjectResponse.data?.chains ?? []
+
+  useEffect(() => {
+    clear()
+  }, [])
+
+  useEffect(() => {
+    return () => {
+      // Cleanup on unmount
+      abortControllerRef.current?.abort()
+    }
+  }, [])
+
+  useEffect(() => {
+    if (outputRef.current) {
+      outputRef.current.scrollTop = outputRef.current.scrollHeight
+    }
+  }, [output])
+
+  useEffect(() => {
+    if (chains[0] && !discoveryChain) {
+      setDiscoveryChain(chains[0].name)
+    }
+  }, [chains, discoveryChain])
+
+  async function handleExecute() {
+    if (!project || !discoveryChain || isRunning) return
+    clear()
+    setIsRunning(true)
+
+    try {
+      const eventSource = executeCommand('discover', project, discoveryChain)
+      abortControllerRef.current = new AbortController()
+
+      eventSource.onmessage = (event) => {
+        const text = event.data.toString()
+        addOutput(text.endsWith('\n') ? text : text + '\n')
+      }
+
+      eventSource.onerror = () => {
+        eventSource.close()
+        setIsRunning(false)
+        queryClient.invalidateQueries({ queryKey: ['projects', project] })
+      }
```

Comment:
onerror fires also on onclose?

Reply:
This is a known quirk of the SSE protocol - normal completion and errors both go through the error handler.

Added a comment.,